### ISSUE TITLE
ROMIO:ad_gpfs hang in collective write due to inconsistent state of gpfsmpio_tuneblocking

### ISSUE THREAD

**Issue by ashwinraghu** [pmodels/mpich#5604]
A user reported a hang when running an application that does collective MPI-IO on a GPFS file system.
On reproducing the issue, I can see that the ranks are waiting on 2 different collectives in their execution path, which shouldn't have been. On examining why, it appears the global gpfsmpio_tuneblocking seems to be inconsistent across the ranks.

From a run on 1 node with 4 ranks:

Rank with lowest PID(probably the aggregator) is stuck here because gpfsmpio_tuneblocking is 1.
#11 0x00002b9cd9ab3d07 in PMPI_Alltoallv () from /lus/cls01075/home/users/ashwinr/mpich-3.4.2/lib/.libs/libmpi.so.12
#12 0x00002b9cda2fc9fb in ADIOI_GPFS_Calc_others_req () from /lus/cls01075/home/users/ashwinr/mpich-3.4.2/lib/.libs/libmpi.so.12
#13 0x00002b9cda3037dd in ADIOI_GPFS_WriteStridedColl () from /lus/cls01075/home/users/ashwinr/mpich-3.4.2/lib/.libs/libmpi.so.12
#14 0x00002b9cda2f9ec0 in MPIOI_File_write_all () from /lus/cls01075/home/users/ashwinr/mpich-3.4.2/lib/.libs/libmpi.so.12
#15 0x00002b9cda2f9fc3 in PMPI_File_write_all () from /lus/cls01075/home/users/ashwinr/mpich-3.4.2/lib/.libs/libmpi.so.12
#16 0x0000000000402367 in main ()
(gdb) frame 13
#13 0x00002b9cda3037dd in ADIOI_GPFS_WriteStridedColl () from /lus/cls01075/home/users/ashwinr/mpich-3.4.2/lib/.libs/libmpi.so.12
(gdb) print (int)gpfsmpio_tuneblocking
$1 = 1

The rest of the ranks are all waiting in another collective because their value for the global is 0.

#7  0x00002af19a628467 in PMPI_Allreduce () from /lus/cls01075/home/users/ashwinr/mpich-3.4.2/lib/.libs/libmpi.so.12
#8  0x00002af19ae7aeba in ADIOI_GPFS_WriteStridedColl () from /lus/cls01075/home/users/ashwinr/mpich-3.4.2/lib/.libs/libmpi.so.12
#9  0x00002af19ae70ec0 in MPIOI_File_write_all () from /lus/cls01075/home/users/ashwinr/mpich-3.4.2/lib/.libs/libmpi.so.12
#10 0x00002af19ae70fc3 in PMPI_File_write_all () from /lus/cls01075/home/users/ashwinr/mpich-3.4.2/lib/.libs/libmpi.so.12
#11 0x0000000000402367 in main ()
(gdb) frame 8
#8  0x00002af19ae7aeba in ADIOI_GPFS_WriteStridedColl () from /lus/cls01075/home/users/ashwinr/mpich-3.4.2/lib/.libs/libmpi.so.12
(gdb) print (int)gpfsmpio_tuneblocking
$1 = 0

What seems to be key to reproducing the problem is the MPI_Info_set() in this code snippet. Keys  other than "romio_no_indep_rw" do not seem to cause the hang though. E.g. MPI_Info_set(info, "romio_ds_write", "disable"); appears to work alright.
``
    MPI_Info_create(&info);
    MPI_Info_set(info, "romio_no_indep_rw", "true");
    MPI_File_open(MPI_COMM_WORLD, filename,
                    MPI_MODE_CREATE | MPI_MODE_RDWR, info, &fh);
    MPI_File_set_view(fh, 0, MPI_INT, newtype, "native", info);
    MPI_File_write_all(fh, writebuf, bufcount, MPI_INT, &status);

**Comment by roblatham00** [comment]
You've been able to reproduce the problem which is tremendous.  Is this a pure MPI-IO stack or is some other library involved?   Setting "romio_no_indep_rw" suggests tremendous familiarity with ROMIO.

**Comment by roblatham00** [comment]
dang, there it is.  `ADIOI_GPFS_Open` *also* calls ad_gpfs_get_env_vars.  and yeah, only aggregators will call that one in the deferred open case.

The code has been structured that way for a long long time, but I admit we don't test deferred open aggressively enough.

I'm surprised gpfs_tuneblocking was not set as part of MPI_File_open:  even though the non-aggregators will skip the call to ADIOI_GPFS_Open, *everyone* is supposed to call ADIOI_GPFS_SetInfo

**Comment by roblatham00** [comment]
wow, that's a subtle mess....

Ok, so we identified the two places that ROMIO's gpfs driver consults environment variables:  at open time and at info processing time.    Unfortunately, only on PE or BGQ do we process gpfs specific hints.  Here's ad_gpfs.c:

```
#if defined(BGQPLATFORM) || defined(PEPLATFORM)
    ADIOI_GPFS_SetInfo, /* SetInfo for BlueGene or PE */
#else
    ADIOI_GEN_SetInfo,  /* SetInfo for any platform besides BlueGene or PE */
#endif
```

this is redundant:  any PE or BlueGene specific code in ADIOI_GPFS_SetInfo is ifdef gaurded

(also BlueGene specific code is guarded by the fact that there aren't any Blue Gene machines out there any more....)

I think I have a fix.   Could also clean up GPFS a lot by dropping Blue Gene cases.
